tfx/examples/cifar10/cifar10_utils_native_keras.py

```diff
@@ -0,0 +1,259 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Python source file includes CIFAR10 utils for Keras model.
+
+The utilities in this file are used to build a model with native Keras.
+This module file will be used in Transform and generic Trainer.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from typing import List, Text
+
+import os
+import absl
+import tensorflow as tf
+import tensorflow_transform as tft
+
+from tfx.components.trainer.rewriting import converters
+from tfx.components.trainer.rewriting import rewriter
+from tfx.components.trainer.rewriting import rewriter_factory
+
+from tfx.components.trainer.executor import TrainerFnArgs
+
+# CIFAR10 dataset has 50000 train records, and 10000 test records
+_TRAIN_DATA_SIZE = 50000
+_EVAL_DATA_SIZE = 10000
+_TRAIN_BATCH_SIZE = 64
+_EVAL_BATCH_SIZE = 64
+
+IMAGE_KEY = 'image'
+LABEL_KEY = 'label'
+
+def transformed_name(key):
+  return key + '_xf'
+
+def _gzip_reader_fn(filenames):
+  """Small utility returning a record reader that can read gzip'ed files."""
+  return tf.data.TFRecordDataset(filenames, compression_type='GZIP')
+
+def _get_serve_tf_examples_fn(model):
+  """Returns a function that feeds the input tensor into the model."""
+
+  @tf.function
+  def serve_tf_examples_fn(image_tensor):
+    """Returns the output to be used in the serving signature."""
+    return model(image_tensor)
+
+  return serve_tf_examples_fn
+
+def _input_fn(file_pattern: List[Text],
+              tf_transform_output: tft.TFTransformOutput,
+              batch_size: int = 200) -> tf.data.Dataset:
+  """Generates features and label for tuning/training.
+
+  Args:
+    file_pattern: List of paths or patterns of input tfrecord files.
+    tf_transform_output: A TFTransformOutput.
+    batch_size: representing the number of consecutive elements of returned
+      dataset to combine in a single batch
+
+  Returns:
+    A dataset that contains (features, indices) tuple where features is a
+      dictionary of Tensors, and indices is a single Tensor of label indices.
+  """
+  transformed_feature_spec = (
+      tf_transform_output.transformed_feature_spec().copy())
+
+  dataset = tf.data.experimental.make_batched_features_dataset(
+      file_pattern=file_pattern,
+      batch_size=batch_size,
+      features=transformed_feature_spec,
+      reader=_gzip_reader_fn,
+      label_key=transformed_name(LABEL_KEY))
+
+  return dataset
+
+def _freeze_model_by_percentage(model: tf.keras.Model,
```

Comment:
There is also a model.trainable parameter lets make sure that it is not overriding the settings here. Can you please verify?

Reply:
I think model.trainable basically set all layers' "trainable" attribute to true, which is equal to _freeze_model_by_percentage(0.0).

Reply:
model.trainable is like a meta switch that turns all parameters trainable or not. if we freeze some layers in the model and then call model.trainable=True, all layers will be unfrozen; if we call model.trainable=False after freezing some of layers, all layers will be frozen.

Reply:
Makes sense. The question is what is the default and does it get in the way.

Reply:
The default for model.trainable is True. As long as we don't modify it it will not get in the way.